op-node/rollup/derive/calldata_source.go

```diff
@@ -54,26 +55,47 @@ type DataSource struct {
 	fetcher L1TransactionFetcher
 	log     log.Logger
 
-	batcherAddr common.Address
+	batcherAddr        common.Address
+	batcherHashVersion uint8
 }
 
 // NewDataSource creates a new calldata source. It suppresses errors in fetching the L1 block if they occur.
 // If there is an error, it will attempt to fetch the result on the next call to `Next`.
-func NewDataSource(ctx context.Context, log log.Logger, cfg *rollup.Config, fetcher L1TransactionFetcher, block eth.BlockID, batcherAddr common.Address) DataIter {
+func NewDataSource(ctx context.Context, log log.Logger, cfg *rollup.Config, fetcher L1TransactionFetcher, block eth.BlockID, batcherAddr common.Address, batcherHashVersion uint8) DataIter {
 	_, txs, err := fetcher.InfoAndTxsByHash(ctx, block.Hash)
 	if err != nil {
 		return &DataSource{
-			open:        false,
-			id:          block,
-			cfg:         cfg,
-			fetcher:     fetcher,
-			log:         log,
-			batcherAddr: batcherAddr,
+			open:               false,
+			id:                 block,
+			cfg:                cfg,
+			fetcher:            fetcher,
+			log:                log,
+			batcherAddr:        batcherAddr,
+			batcherHashVersion: batcherHashVersion,
 		}
-	} else {
+	}
+	_, receipts, err := fetcher.FetchReceipts(ctx, block.Hash)
+	if err != nil {
 		return &DataSource{
-			open: true,
-			data: DataFromEVMTransactions(cfg, batcherAddr, txs, log.New("origin", block)),
+			open:               false,
+			id:                 block,
+			cfg:                cfg,
+			fetcher:            fetcher,
+			log:                log,
+			batcherAddr:        batcherAddr,
+			batcherHashVersion: batcherHashVersion,
+		}
+	} else {
+		if batcherHashVersion == 0 {
+			return &DataSource{
+				open: true,
+				data: DataFromEVMTransactions(cfg, batcherAddr, txs, log.New("origin", block)),
+			}
+		} else {
+			return &DataSource{
+				open: true,
+				data: DataFromEVMTransactionsV2(txs, receipts, log.New("origin", block)),
```

Comment:
Why don't we need `cfg` as an argument to the `DataFromEVMTransactionsV2` function?

Reply:
Because the `V2` function does not access any of it's fields anymore.

Reply:
Hmm. Maybe the V2 version still needs to check that `to` matches the batch inbox address. Think I have to look a bit more into where the transactions come from to see if they are already filtered by destination address or not.

Reply:
Yeah it looks like it iterates through all the L1 transactions so I think we need to filter out the ones that aren't sent to the batch inbox contract address. Thanks.

Reply:
https://github.com/EspressoSystems/op-leader-election/pull/23/commits/5e508849f65372268161033eb1e151d2fc732a76